### PR TITLE
add basic zone-create functionality

### PR DESCRIPTION
**Pull request by mdb**
This adds basic `zone-create` functionality.

### CODE REVIEW

File: vinyldns.go [hunk 1]
```diff
@@ -142,6 +142,58 @@ func main() {
...
+				cli.StringFlag{
+					Name:  "email",
+					Usage: "The zone email",
+				},
+				cli.StringFlag{
+					Name:  "admin-group-id",
+					Usage: "The zone admin group ID",
+				},
+				cli.StringFlag{
+					Name:  "transfer-connection-name",
```
Comment (pauljamescleary) [inline vinyldns.go]: I know technically we have this `connection name` thing, but it is not used in practice for anything.  I wonder to simplify things slightly if we omit it and just force it to be the same value as the key name.

File: vinyldns.go [hunk 2]
```diff
@@ -142,6 +142,58 @@ func main() {
...
+				cli.StringFlag{
+					Name:  "zone-connection-key-name",
+					Usage: "The zone connection key name",
+				},
+				cli.StringFlag{
+					Name:  "zone-connection-key",
+					Usage: "The zone connection key",
+				},
+				cli.StringFlag{
+					Name:  "zone-connection-primary-server",
```
Comment (pauljamescleary) [inline vinyldns.go]: on thing missing from the zone info is the ACL rules.  That is probably a ton though from a CLI standpoint.  Worth noting, but I wouldn't add it here.  Fact is, ACL rules are not widely used.

File: vinyldns.go [hunk 3]
```diff
@@ -142,6 +142,58 @@ func main() {
...
+			Flags: []cli.Flag{
+				cli.StringFlag{
+					Name:  "name",
+					Usage: "The zone name",
+				},
+				cli.StringFlag{
+					Name:  "email",
+					Usage: "The zone email",
+				},
+				cli.StringFlag{
```
Comment (pauljamescleary) [inline vinyldns.go]: would it be easier from a UX perspective to take an admin group name instead of id?

File: vinyldns.go [hunk 4]
```diff
@@ -142,6 +142,58 @@ func main() {
...
+					Name:  "admin-group-id",
+					Usage: "The zone admin group ID",
+				},
+				cli.StringFlag{
+					Name:  "transfer-connection-name",
+					Usage: "The zone transfer connection",
+				},
+				cli.StringFlag{
```
Comment (pauljamescleary) [inline vinyldns.go]: should note that the connection and transfer connection sets of fields are optional.  We assume a global default.  Most users do not (and really should not) have their own connections.

As long as the cli does not mandate the fields, we are good.

Comment (mdb) [reply vinyldns.go]: Gotcha; thanks for the info.

File: vinyldns.go (hunk 3 above)
Comment (mdb) [reply vinyldns.go]: Yes, arguably. Will the API accept that, or would that UX require I...
1. call the API to find the group ID of the group with the provided name
2. use the group ID from step #1 in the zone creation call
?

File: vinyldns.go (hunk 2 above)
Comment (mdb) [reply vinyldns.go]: Are ACL rules declared during zone creation? Or are they something that are only used in reading the zone?

File: vinyldns.go (hunk 1 above)
Comment (mdb) [reply vinyldns.go]: Are you saying that `Name` and `KeyName` are the same thing and therefore redundant on transfer and zone connections?

Comment (pauljamescleary) [reply vinyldns.go]: `Name` is just not used, meaningless from an app standpoint.  Really should be deprecated / deleted in a future version of vinyldns.

File: vinyldns.go (hunk 2 above)
Comment (pauljamescleary) [reply vinyldns.go]: They _can_ be declared on zone creation.  I would argue though that they are really cumbersome from a cli / ux perspective and should be omitted.

File: vinyldns.go (hunk 3 above)
Comment (pauljamescleary) [reply vinyldns.go]: You would have to the steps you mentioned.

It also provides some nice communication.  If the caller enters a name for a group they do not belong to, we can predictably fail with a nice error message :)

File: vinyldns.go (hunk 1 above)
Comment (mdb) [reply vinyldns.go]: Awesome; I'll remove it.

File: vinyldns.go (hunk 2 above)
Comment (mdb) [reply vinyldns.go]: Alrighty; let's punt on ACL rules with regard to zone creation, though I [made an issue](https://github.com/vinyldns/vinyldns-cli/issues/28) surrounding displaying them via the `zone` command.

File: vinyldns.go (hunk 3 above)
Comment (mdb) [reply vinyldns.go]: I like that, though I'm inclined to suggest we proceed in using `id` for now, as we do elsewhere for other resource types. Then, we can roll out a `--admin-group-name` flag in the future. If you agree, I can create an issue to do so.

File: vinyldns.go (hunk 1 above)
Comment (mdb) [reply vinyldns.go]: @pauljamescleary Maybe I'm a bit confused. When I remove `Connection.Name`, I get the following -- did I misinterpret you?

```
 ✗ zone-create
   (in test file tests/vinyldns_tests.bats, line 23)
     `)' failed
   Request URL:
   http://localhost:9000/zones
   Request Method:
   POST
   Request body:
   {"name":"ok.","email":"test@test.com","adminGroupId":"ok-group","connection":{"keyName":"vinyldns.","key":"nzisn+4G2ldMn0q1CV3vsg==","primaryServer":"vinyldns-bind9"},"transferConnection":{"keyName":"vinyldns.","key":"nzisn+4G2ldMn0q1CV3vsg==","primaryServer":"vinyldns-bind9"}}
   Response code:
   400
   Response body:
   {"errors":["Missing ZoneConnection.name"]}
```

Comment (pauljamescleary) [reply vinyldns.go]: Right, if you _do_ put the connection you have to put the connection name in.  Sorry if that wasn't clear.

My comment was, from a cli standpoint, you don't need to ask the user for it.  In your code, just set the connection name to the same value as the keyName.

Comment (mdb) [reply vinyldns.go]: Ah, gotcha. Thanks for clarifying!

**Comment by mdb** [comment]
@pauljamescleary Can you clarify for me why, if zone creation requires no connection, the following occurs?

```
 ✗ zone-create (no connection)
   (in test file tests/vinyldns_tests.bats, line 37)
     `)' failed
   Request URL:
   http://localhost:9000/zones
   Request Method:
   POST
   Request body:
   {"name":"ok-no-connection.","email":"test@test.com","adminGroupId":"ok-group"}
   Response code:
   400
   Response body:
   "Unable to connect to zone: The operation was refused by the server: REFUSED"
```

**Comment by pauljamescleary** [comment]
@mdb This is going to sound like I am contradicting myself.

`REFUSED` is a message coming back from the backend DNS server itself, not vinyldns.

VinylDNS _connects to existing zones_, it does not "create" zones.  So, the zone that you are connecting to must already exist in the backend bind9 server.

The zone you are connecting to `"ok-no-connection."` in your example does not exist in our default bind9 test setup.

Try using the zone name "vinyldns." instead and let me know how you make out.

**Comment by nimaeskandary** [comment]
For reference, these are the zones you can work with out of the box with the bind9 docker image: https://github.com/vinyldns/vinyldns/tree/master/docker/bind9/zones

File: vinyldns.go (hunk 3 above)
Comment (pauljamescleary) [reply vinyldns.go]: @mdb sounds good

**Comment by mdb** [comment]
@pauljamescleary Yes, I believe the connections are now optional. However, I don't yet have test cases asserting so. Would you mind providing me details for a zone creation request that should work with no connection? I just want to be sure I have a valid test case.

**Pull request merged**